src/scanner/mod.rs

```diff
@@ -100,41 +103,59 @@ impl Scanner {
+        let mut ret = Err(io::Error::new(
+            io::ErrorKind::Other,
+            format!("Did not connect to {}", &socket),
+        ));
+
+        for nr_try in 0..self.tries {
+            debug!("Try number: {}", nr_try + 1);
+
+            ret = match self.connect(socket).await {
+                Ok(x) => {
+                    debug!(
+                        "Connection was successful, shutting down stream {}",
+                        &socket
+                    );
+                    match x.shutdown(Shutdown::Both) {
+                        Err(e) => debug!("Shutdown stream error {}", &e),
+                        _ => {}
                     }
+                    if !self.greppable {
+                        if self.accessible {
+                            println!("Open {}", socket.to_string());
+                        } else {
+                            println!("Open {}", socket.to_string().purple());
+                        }
+                    }
+
+                    Ok(socket)
                 }
+                Err(e) => match e.kind() {
+                    ErrorKind::Other => {
+                        if e.to_string().contains("No route to host")
+                            || e.to_string().contains("Network is unreachable")
+                        {
+                            debug!("Socket connect error: {} {}", &e.to_string(), &socket);
+                            Err(io::Error::new(io::ErrorKind::Other, e.to_string()))
+                        } else {
+                            debug!("Socket connect error: {} {}", &e.to_string(), &socket);
+                            panic!("Too many open files. Please reduce batch size. The default is 5000. Try -b 2500.");
+                        }
+                    }
+                    _ => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
+                },
+            };
 
-                Ok(socket)
+            if ret.is_ok() {
+                debug!("Return ok after {} tries", nr_try + 1);
+                return ret;
             }
-            Err(e) => match e.kind() {
-                ErrorKind::Other => {
-                    if e.to_string().contains("No route to host")
-                        || e.to_string().contains("Network is unreachable")
-                    {
-                        debug!("Socket connect error: {} {}", &e.to_string(), &socket);
-                        Err(io::Error::new(io::ErrorKind::Other, e.to_string()))
-                    } else {
-                        debug!("Socket connect error: {} {}", &e.to_string(), &socket);
-                        panic!("Too many open files. Please reduce batch size. The default is 5000. Try -b 2500.");
-                    }
-                }
-                _ => Err(io::Error::new(io::ErrorKind::Other, e.to_string())),
-            },
         }
+
+        debug!("Return after {} tries", self.tries);
+
+        return ret;
```

Comment:
The value of `ret` here will always be the error of the last iteration, because all other values will be overridden again or discarded if a connection succeeds. I think it would be clearer, if you checked whether there will be another iteration in the loop body and if not return the current (and therefore the last) error.

Reply:
If I understand correctly you proposed to write something like:
```Rust
Err(e) => {
  match e.kind() {
    ...
  }
  if nr_try == self.tries {
    return e;
  }
}
```
I do not really stand for this solution because it adds an `if` inside the loop which already contains an `if`.
I rather write something like this:
```Rust
for ... {
  ... {
    Ok(x) => {
    }
    Err(e) => {
      ret = e;
      match e.kind() => {
        ...
      }
    }
  }
}
return ret;
```
Since the code after the loop means the loop is finished an so code tries `self.tries` which is a problem.

Reply:
Okay sure, this is probably more about style than anything else. As I see it, there is a local variable that gets reassigned in every iteration of the loop and every value except for the last one is discarded in the next iteration anyway, so there is no reason to keep any of them around. Logically, the function returns either an `Ok` with the socket, or it returns the error message from the last try. I think that the `if` versions better reflects this logic, as the last error directly comes from the last connection attempt and is returned without another indirection. With the `if`, the code that handles connection errors and translates them into the API of the function is in one place and it is more clear what is happening.

Although I think that the extra `if` adds less complexity than the extra variable, I could see that one would be unhappy with the code. If the `ret` variable sticks around, here is what I propose: I think something like `last_error` or so would be a more fitting name, because that's what it is effectively. Furthermore, the type of it should IMO be `Option<io::Error>` (initially `None`), so that you don't have to allocate the default error message if you go into the loop. Lastly, the assignment should move into the `Err` match arm (as you already did in the example above), because currently `ret` only holds `Err` variants anyway.

Sorry for the big comment, but I think the current code around `ret` really is more complicated than it has to be. That said, I would be fine with the second approach if other reviewers give it a pass.

Reply:
I will go with the `if` inside the loop.
I do not think adding a branch will have a great impact on performance and I agree on the fact that the code is more readable with it.